contrib/qpl-cmake/CMakeLists.txt

```diff
@@ -0,0 +1,362 @@
+               ${QPL_SRC_DIR}/isal/igzip/huff_codes.c
+               ${QPL_SRC_DIR}/isal/igzip/hufftables_c.c
+               ${QPL_SRC_DIR}/isal/igzip/igzip.c
+               ${QPL_SRC_DIR}/isal/igzip/igzip_base.c
+               ${QPL_SRC_DIR}/isal/igzip/flatten_ll.c
+               ${QPL_SRC_DIR}/isal/igzip/encode_df.c
+               ${QPL_SRC_DIR}/isal/igzip/igzip_icf_base.c
+               ${QPL_SRC_DIR}/isal/igzip/igzip_inflate.c
+               ${QPL_SRC_DIR}/isal/igzip/igzip_icf_body.c
+               ${QPL_SRC_DIR}/isal/crc/crc_base.c
+               ${QPL_SRC_DIR}/isal/crc/crc64_base.c)
+
+set(ISAL_ASM_SRC ${QPL_SRC_DIR}/isal/igzip/igzip_body.asm
+                 ${QPL_SRC_DIR}/isal/igzip/igzip_gen_icf_map_lh1_04.asm
+                 ${QPL_SRC_DIR}/isal/igzip/igzip_gen_icf_map_lh1_06.asm
+                 ${QPL_SRC_DIR}/isal/igzip/igzip_decode_block_stateless_04.asm
+                 ${QPL_SRC_DIR}/isal/igzip/igzip_finish.asm
+                 ${QPL_SRC_DIR}/isal/igzip/encode_df_04.asm
+                 ${QPL_SRC_DIR}/isal/igzip/encode_df_06.asm
+                 ${QPL_SRC_DIR}/isal/igzip/igzip_decode_block_stateless_01.asm
+                 ${QPL_SRC_DIR}/isal/igzip/proc_heap.asm
+                 ${QPL_SRC_DIR}/isal/igzip/igzip_icf_body_h1_gr_bt.asm
+                 ${QPL_SRC_DIR}/isal/igzip/igzip_icf_finish.asm
+                 ${QPL_SRC_DIR}/isal/igzip/igzip_inflate_multibinary.asm
+                 ${QPL_SRC_DIR}/isal/igzip/igzip_update_histogram_01.asm
+                 ${QPL_SRC_DIR}/isal/igzip/igzip_update_histogram_04.asm
+                 ${QPL_SRC_DIR}/isal/igzip/rfc1951_lookup.asm
+                 ${QPL_SRC_DIR}/isal/igzip/adler32_sse.asm
+                 ${QPL_SRC_DIR}/isal/igzip/adler32_avx2_4.asm
+                 ${QPL_SRC_DIR}/isal/igzip/igzip_deflate_hash.asm
+                 ${QPL_SRC_DIR}/isal/igzip/igzip_set_long_icf_fg_04.asm
+                 ${QPL_SRC_DIR}/isal/igzip/igzip_set_long_icf_fg_06.asm
+                 ${QPL_SRC_DIR}/isal/igzip/igzip_multibinary.asm
+                 ${QPL_SRC_DIR}/isal/igzip/stdmac.asm
+                 ${QPL_SRC_DIR}/isal/crc/crc_multibinary.asm
+                 ${QPL_SRC_DIR}/isal/crc/crc32_gzip_refl_by8.asm
+                 ${QPL_SRC_DIR}/isal/crc/crc32_gzip_refl_by8_02.asm
+                 ${QPL_SRC_DIR}/isal/crc/crc32_gzip_refl_by16_10.asm
+                 ${QPL_SRC_DIR}/isal/crc/crc32_ieee_01.asm
+                 ${QPL_SRC_DIR}/isal/crc/crc32_ieee_02.asm
+                 ${QPL_SRC_DIR}/isal/crc/crc32_ieee_by4.asm
+                 ${QPL_SRC_DIR}/isal/crc/crc32_ieee_by16_10.asm
+                 ${QPL_SRC_DIR}/isal/crc/crc32_iscsi_00.asm
+                 ${QPL_SRC_DIR}/isal/crc/crc32_iscsi_01.asm
+                 ${QPL_SRC_DIR}/isal/crc/crc32_iscsi_by16_10.asm)
+
+set(DISABLED_WARNINGS_FLAGS /wd4244 /wd4101 /wd4267 /wd4018 /wd4146 /wd4334 /wd4308)
+set(FEATURE_FLAGS -DHAVE_AS_KNOWS_AVX512 -DAS_FEATURE_LEVEL=10)
```

Comment:
AS_FEATURE_LEVEL=10 means "Check SIMD capabilities of the target system at runtime and use up to AVX512 if available"
HAVE_KNOWS_AVX512 means rely on AVX512 being available on the target system. I'll make it dependent on ENABLE_AVX512.
According to QPIs fallback path requirement, I've added commits to make sure ENABLE_QPL dependent on ENABLE_AVX2/ENABLE_AVX512.

Reply:
Having your comment about AS_FEATURE_LEVEL also in the CMakeLists would be nice :D

Reply:
About minimum AVX requirements: I think the best solution is if
- QPL is completely independent of the SIMD-level compiler flags which ClickHouse is compiled with (which are in turn controlled by ENABLE_SSE42/ENABLE_AVX/ENABLE_AVX2 etc.) and, at the same time,
- checks at runtime the SSE/AVX capabilities of the target system.

Setting ENABLE_AVX512, as done right now in cmake/cpu_features.cmake, is unfortunately not an option as I mentioned in my above comment. It would break ClickHouse on all non-recent systems.

Even if we go with ideal solution above: The SW fallback requires AVX1, right? So we need to raise the minimum SIMD version to that, i.e. set ENABLE_AVX = 1. Since this could have some performance implications on its own, we should do that in a separate PR. https://github.com/ClickHouse/ClickHouse/pull/34069 was an earlier attempt, we can try to resurrect it.

Reply:
The SW fallback requires AVX2, not AVX1. As you mentioned,  "raise the minimum SIMD version to AVX1" is on evaluation but AVX2 is too aggressive. So I think currently QPL cannot be independent of the Clickhouse compiler 
 flags. I've made QPL dependent on ENABLE_AVX2 OR ENABLE_AVX512, the user who build clickhouse would make decision to enable it or not according to target system.

> if (OS_LINUX AND ARCH_AMD64 AND (ENABLE_AVX2 OR ENABLE_AVX512))
>     option (ENABLE_QPL "Enable Intel® Query Processing Library" ${ENABLE_LIBRARIES})

Reply:
> Having your comment about AS_FEATURE_LEVEL also in the CMakeLists would be nice :D

Done

Reply:
Okay then, thanks for the clarification.